README.md
# Percona MongoDB Exporter

[![Release](https://img.shields.io/github/release/percona/mongodb_exporter.svg?style=flat)](https://github.com/ghostbaby/mongodb_exporter/releases/latest)
[![Build Status](https://travis-ci.com/percona/mongodb_exporter.svg?branch=master)](https://travis-ci.com/percona/mongodb_exporter)
[![codecov.io Code Coverage](https://img.shields.io/codecov/c/github/percona/mongodb_exporter.svg?maxAge=2592000)](https://codecov.io/github/percona/mongodb_exporter?branch=master)
[![Go Report Card](https://goreportcard.com/badge/github.com/ghostbaby/mongodb_exporter)](https://goreportcard.com/report/github.com/ghostbaby/mongodb_exporter)
[![CLA assistant](https://cla-assistant.percona.com/readme/badge/percona/mongodb_exporter)](https://cla-assistant.percona.com/percona/mongodb_exporter)

Based on [MongoDB exporter](https://github.com/dcu/mongodb_exporter) by David Cuadrado ([@dcu](https://github.com/dcu)), but forked for full sharded support and structure changes.

## Features

- MongoDB Server Status metrics (*cursors, operations, indexes, storage, etc*)
- MongoDB Replica Set metrics (*members, ping, replication lag, etc*)
- MongoDB Replication Oplog metrics (*size, length in time, etc*)
- MongoDB Sharding metrics (*shards, chunks, db/collections, balancer operations*)
- MongoDB RocksDB storage-engine metrics (*levels, compactions, cache usage, i/o rates, etc*)
- MongoDB WiredTiger storage-engine metrics (*cache, blockmanger, tickets, etc*)
- MongoDB Top Metrics per collection (writeLock, readLock, query, etc*)

### Important Note

Metrics `mongodb_mongod_replset_oplog_*` doesn't work in [Master/Slave](https://docs.mongodb.com/v3.4/core/master-slave/) replication mode, because it was *DEPRECATED* in MongoDB `3.2` and removed in `4.0`.

## Building and running

### Building

Just run `make`. It will install all needed tools, format code with `go fmt`, build a binary for your OS inside `./dist` directory and runs tests.

*Note: You need to have `docker` installed to run tests as it uses mongodb. Also port `27017` must been freed as `docker-compos` maps this port into your host OS while testing.*

```bash
make
```

If you want just build a binary for your OS without codestyle checks and tests you can run command below:

```bash
make build
```

If you don't have or don't want to install the whole GO stuff, use this docker build that creates a container with a freshly built `mongodb_exporter` binary:

```bash
make docker
```

### Running

To define your own MongoDB URL, use environment variable `MONGODB_URI`. If set this variable takes precedence over `--mongodb.uri` flag.

To enable HTTP basic authentication, set environment variable `HTTP_AUTH` to user:password pair. Alternatively, you can
use YAML file with `server_user` and `server_password` fields.

```bash
export MONGODB_URI='mongodb://localhost:27017'
export HTTP_AUTH='user:password'
./bin/mongodb_exporter [<flags>]
```

#### Kubernetes

You can use the chart [prometheus-mongodb-exporter](https://github.com/helm/charts/tree/master/stable/prometheus-mongodb-exporter) from helm stable repository.

### Flags

See the help page with `-h`.

If you use [MongoDB Authorization](https://docs.mongodb.org/manual/core/authorization/), you must:

1. Create a user with '*clusterMonitor*' role and '*read*' on the '*local*' database, like the following (*replace username/password!*):

```js
db.getSiblingDB("admin").createUser({
    user: "mongodb_exporter",
    pwd: "s3cr3tpassw0rd",
    roles: [
        { role: "clusterMonitor", db: "admin" },
        { role: "read", db: "local" }
    ]
})
```

2. Set environment variable `MONGODB_URI` before starting the exporter:

```bash
export MONGODB_URI=mongodb://mongodb_exporter:s3cr3tpassw0rd@localhost:27017
```

If you use [x.509 Certificates to Authenticate Clients](https://docs.mongodb.com/manual/tutorial/configure-x509-client-authentication/), pass in username and `authMechanism` via [connection options](https://docs.mongodb.com/manual/reference/connection-string/#connections-connection-options) to the MongoDB uri. Eg:

```
mongodb://CN=myName,OU=myOrgUnit,O=myOrg,L=myLocality,ST=myState,C=myCountry@localhost:27017/?authMechanism=MONGODB-X509
```

## Note about how this works

Point the process to any mongo port and it will detect if it is a mongos, replicaset member, or stand alone mongod and return the appropriate metrics for that type of node. This was done to prevent the need to an exporter per type of process.

## Roadmap

- Document more configurations options here
- Stabilize RocksDB and WiredTiger support
- Move MongoDB user/password/authdb to a file (for security)
- Write more go tests


## Submitting Bug Reports

If you find a bug in Percona MongoDB Exporter or one of the related projects, you should submit a report to that project's [JIRA](https://jira.percona.com) issue tracker.

Your first step should be [to search](https://jira.percona.com/issues/?jql=project=PMM%20AND%20component=MongoDB_Exporter) the existing set of open tickets for a similar report. If you find that someone else has already reported your problem, then you can upvote that report to increase its visibility.

If there is no existing report, submit a report following these steps:

1. [Sign in to Percona JIRA.](https://jira.percona.com/login.jsp) You will need to create an account if you do not have one.
2. [Go to the Create Issue screen and select the relevant project.](https://jira.percona.com/secure/CreateIssueDetails!init.jspa?pid=11600&issuetype=1&priority=3&components=11603)
3. Fill in the fields of Summary, Description, Steps To Reproduce, and Affects Version to the best you can. If the bug corresponds to a crash, attach the stack trace from the logs.

An excellent resource is [Elika Etemad's article on filing good bug reports.](http://fantasai.inkedblade.net/style/talks/filing-good-bugs/).

As a general rule of thumb, please try to create bug reports that are:

- *Reproducible.* Include steps to reproduce the problem.
- *Specific.* Include as much detail as possible: which version, what environment, etc.
- *Unique.* Do not duplicate existing tickets.
- *Scoped to a Single Bug.* One bug per report.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Changed

### Added

### Fixed

## [0.8.0]
### Changed
- [PMM-3511](https://jira.percona.com/browse/PMM-3511): Switched to [mongo-go-driver](https://github.com/mongodb/mongo-go-driver).
- [PMM-4168](https://jira.percona.com/browse/PMM-4168): Updated all dependencies including `github.com/prometheus/golang_client`.
- Moved `opcounters`, `opcountersRepl` to `common_collector` (#146), thx [@nikita-b](https://github.com/nikita-b).

### Added

### Fixed

## [0.7.1]
### Added
- Added Authentification Database option when connect to mongo #139, thx [@etiennecoutaud](https://github.com/etiennecoutaud).
- Added helm chart to readme #140, thx [@pgdagenais](https://github.com/pgdagenais).
- [PMM-4154](https://jira.percona.com/browse/PMM-4154): Added standard logging flags.

### Fixed
- Fixed some function comments based on best practices from Effective Go #137, thx [@CodeLingoBot](https://github.com/CodeLingoBot).
- [PMM-3473](https://jira.percona.com/browse/PMM-3473): Fixed panic and runtime error.

## [0.7.0]
### Changed
- [PMM-3512](https://jira.percona.com/browse/PMM-3512): Switched to [kingpin](https://github.com/alecthomas/kingpin) library.
This is a **BREAKING CHANGE** because kingpin uses `--` instead of `-` for long flags, so be careful when updating.
- [PMM-2261](https://jira.percona.com/browse/PMM-2261) Unify common mongod and mongos server metrics, thx [@bz2](https://github.com/bz2)
This is a **BREAKING CHANGE**. The labels of these metrics are now prefixed with just `mongodb_` rather than `mongodb_mongo[ds]_`.

### Added
- Fine grained error handling for index usage and collection stats (#128), thx [@akira-kurogane](https://github.com/akira-kurogane)
- Introduce a docker go build that creates a mongodb_exporter binary within a container (#112), thx [@mminks](https://github.com/mminks)
- Ability to make releases and snapshots with [GoReleaser](https://goreleaser.com/)

## [0.6.3] - 2019-02-13
### Added
- PMM-3401: Added collection of TTL metrics #127, thx [@fastest963](https://github.com/fastest963)
- Added some new metrics:
  - `member_replication_lag`
  - `member_operational_lag`
  - `op_latencies_latency_total`
  - `op_latencies_ops_total`
  - `op_latencies_histogram`

### Fixed
- Fix for broken labels on `index_usage` metrics
- Fix SIGSEGV when running connected to a Primary, thx [@anas-aso](https://github.com/anas-aso)

### Changed
- Move CHANGELOG.md to "Keep a Changelog" format.

## [0.6.2] - 2018-09-11
### Added
- Build binaries #110
- Test `--help` flag for diff of options between releases #111

## [0.6.1] - 2018-06-15
### Fixed
- `--version` now properly reports `0.6.1`

## [0.6.0] - 2018-06-13
### Added
- Add required testify dependency #107, thx [@RubenHoms](https://github.com/RubenHoms)
- Add timeout flags #100, thx [@unguiculus](https://github.com/unguiculus)
- Enable collection of table top metrics #94, thx [@bobera](https://github.com/bobera)
- Individual index usage stats and index sizes added #97, thx [@martinhoefling](https://github.com/martinhoefling)
- Support `&ssl=true` #105 #90, thx [@dbmurphy](https://github.com/dbmurphy)

### Fixed
- Fix `balancerIsEnabled` & `balancerChunksBalanced` values #106, thx [@jmsantorum](https://github.com/jmsantorum)

## [0.5.0] - 2018-05-24
### Added
- Check connection with exporter #92. Adds `--test` flag to verify connection with MongoDB and quits.

### Removed
- Removed tests for EOL'ed MongoDB 3.0 and Percona Server for MongoDB 3.0.

### Fixed
- Redact Mongo URI #101. Fixes URI logging in plain text including credentials when no session can be created.
- ARM64-specific fixes #102. Fixed two portability issues in Makefile.

## [0.4.0] - 2018-01-17
### Added
- New flags `-collect.database` and `-collect.collection` can be used to enable collection of database and collection
  metrics. They are disabled by default.
- MongoDB connections are now kept between the scrapes. New flag `-mongodb.max-connections` (with the default value `1`)
controls the maximum number of established connections.
- Add standard metrics:
  - `mongodb_scrape_errors_total`
  - `mongodb_up`
- Some queries now contain [cursor comments](https://www.percona.com/blog/2017/06/21/tracing-mongodb-queries-to-code-with-cursor-comments/)
with source code locations.

### Changed
- Go vendoring switched to [dep](https://github.com/golang/dep).

## [0.3.1] - 2017-09-08
### Changed
- Better logging for scrape errors.

## [0.3.0] - 2017-07-07
### Added
- Add standard metrics:
  - `mongodb_exporter_scrapes_total`
  - `mongodb_exporter_last_scrape_error`
  - `mongodb_exporter_last_scrape_duration_seconds`

### Fixed
- Fix a few data races.

## [0.2.0] - 2017-06-28
### Changed
- Default listen port changed to 9216.
- All log messages now go to stderr. Logging flags changed.
- Fewer messages on default INFO logging level.
- Use https://github.com/prometheus/common log for logging instead of https://github.com/golang/glog.
- Use https://github.com/prometheus/common version to build with version information.
- Use https://github.com/prometheus/promu for building.

## [0.1.0] - 2016-04-13
### Added
- First tagged version.

[Unreleased]: https://github.com/ghostbaby/mongodb_exporter/compare/v0.8.0...HEAD
[0.8.0]: https://github.com/ghostbaby/mongodb_exporter/compare/v0.7.1...v0.8.0
[0.7.1]: https://github.com/ghostbaby/mongodb_exporter/compare/v0.7.0...v0.7.1
[0.7.0]: https://github.com/ghostbaby/mongodb_exporter/compare/v0.6.3...v0.7.0
[0.6.3]: https://github.com/ghostbaby/mongodb_exporter/compare/v0.6.2...v0.6.3
[0.6.2]: https://github.com/ghostbaby/mongodb_exporter/compare/v0.6.1...v0.6.2
[0.6.1]: https://github.com/ghostbaby/mongodb_exporter/compare/v0.6.0...v0.6.1

mongodb_exporter_test.go
package main

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"net"
	"net/http"
	"net/url"
	"os"
	"os/exec"
	"path"
	"reflect"
	"regexp"
	"runtime"
	"strconv"
	"strings"
	"syscall"
	"testing"
	"time"

	"github.com/andreyvit/diff"
	pmmVersion "github.com/percona/pmm/version"
	"github.com/prometheus/common/version"
	"github.com/stretchr/testify/assert"

	"github.com/ghostbaby/mongodb_exporter/shared"
)

var Update = flag.Bool("update", false, "update .golden files")

// bin stores information about path of executable and attached port
type bin struct {
	path string
	port int
}

// TestBin builds, runs and tests binary.
func TestBin(t *testing.T) {
	if testing.Short() {
		t.Skip("-short is passed, skipping integration test")
	}

	var err error
	binName := "mongodb_exporter"

	binDir, err := ioutil.TempDir("/tmp", binName+"-test-bindir-")
	if err != nil {
		t.Fatal(err)
	}
	defer func() {
		err := os.RemoveAll(binDir)
		if err != nil {
			t.Fatal(err)
		}
	}()

	importpath := "github.com/ghostbaby/mongodb_exporter/vendor/github.com/percona/pmm"
	path := binDir + "/" + binName
	xVariables := map[string]string{
		importpath + "/version.Version":    "gotest-version",
		importpath + "/version.Branch":     "gotest-branch",
		importpath + "/version.FullCommit": "gotest-revision",
	}
	var ldflags []string
	for x, value := range xVariables {
		ldflags = append(ldflags, fmt.Sprintf("-X %s=%s", x, value))
	}
	cmd := exec.Command(
		"go",
		"build",
		"-o",
		path,
		"-ldflags",
		strings.Join(ldflags, " "),
	)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	err = cmd.Run()
	if err != nil {
		t.Fatalf("Failed to build: %s", err)
	}

	tests := []func(*testing.T, bin){
		testFlagHelp,
		testFlagTest,
		testFlagTestWithTLS,
		testFlagVersion,
		testLandingPage,
		testDefaultGatherer,
		testBuildVersionGatherer,
	}

	portStart := 56000
	t.Run(binName, func(t *testing.T) {
		for _, f := range tests {
			f := f // capture range variable
			fName := runtime.FuncForPC(reflect.ValueOf(f).Pointer()).Name()
			portStart++
			data := bin{
				path: path,
				port: portStart,
			}
			t.Run(fName, func(t *testing.T) {
				t.Parallel()
				f(t, data)
			})
		}
	})
}

func testFlagHelp(t *testing.T, data bin) {
	cmd := exec.Command(
		data.path,
		"--help",
	)

	output, _ := cmd.CombinedOutput()
	output = regexp.MustCompile(regexp.QuoteMeta(data.path)).ReplaceAll(output, []byte("mongodb_exporter"))
	actual := string(output)

	filename := path.Join("testdata", path.Base(t.Name())+".golden")
	if *Update {
		err := ioutil.WriteFile(filename, output, 0600)
		assert.NoError(t, err)
	}
	b, err := ioutil.ReadFile(filename)
	assert.NoError(t, err)
	expected := string(b)

	if actual != expected {
		t.Errorf("diff:\n%s", diff.LineDiff(expected, actual))
	}
}

func testFlagVersion(t *testing.T, data bin) {
	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
	defer cancel()

	cmd := exec.CommandContext(
		ctx,
		data.path,
		"--version",
		"--web.listen-address", fmt.Sprintf(":%d", data.port),
	)

	b := &bytes.Buffer{}
	cmd.Stdout = b
	cmd.Stderr = b

	if err := cmd.Run(); err != nil {
		t.Fatal(err)
	}

	expectedRegexp := `mongodb_exporter, version gotest-version \(branch: gotest-branch, revision: gotest-revision\)
  build user:
  build date:
  go version:
`

	expectedScanner := bufio.NewScanner(bytes.NewBufferString(expectedRegexp))
	defer func() {
		if err := expectedScanner.Err(); err != nil {
			t.Fatal(err)
		}
	}()

	gotScanner := bufio.NewScanner(b)
	defer func() {
		if err := gotScanner.Err(); err != nil {
			t.Fatal(err)
		}
	}()

	for gotScanner.Scan() {
		if !expectedScanner.Scan() {
			t.Fatalf("didn't expected more data but got '%s'", gotScanner.Text())
		}
		ok, err := regexp.MatchString(expectedScanner.Text(), gotScanner.Text())
		if err != nil {
			t.Fatal(err)
		}
		if !ok {
			t.Fatalf("'%s' does not match regexp '%s'", gotScanner.Text(), expectedScanner.Text())
		}
	}

	if expectedScanner.Scan() {
		t.Errorf("expected '%s' but didn't got more data", expectedScanner.Text())
	}
}

func testLandingPage(t *testing.T, data bin) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	// Run exporter.
	cmd := exec.CommandContext(
		ctx,
		data.path,
		"--web.listen-address", fmt.Sprintf(":%d", data.port),
	)
	if err := cmd.Start(); err != nil {
		t.Fatal(err)
	}
	defer cmd.Wait()
	defer cmd.Process.Kill()

	// Get the main page.
	urlToGet := fmt.Sprintf("http://127.0.0.1:%d", data.port)
	body, err := waitForBody(urlToGet)
	if err != nil {
		t.Fatal(err)
	}
	got := string(body)

	expected := `<html>
<head>
	<title>MongoDB exporter</title>
</head>
<body>
	<h1>MongoDB exporter</h1>
	<p><a href="/metrics">Metrics</a></p>
</body>
</html>`
	assert.Equal(t, expected, got)
}

func testDefaultGatherer(t *testing.T, data bin) {
	metricPath := "/metrics"
	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()

	cmd := exec.CommandContext(
		ctx,
		data.path,
		"--web.telemetry-path", metricPath,
		"--web.listen-address", fmt.Sprintf(":%d", data.port),
	)

	if err := cmd.Start(); err != nil {
		t.Fatal(err)
	}
	defer cmd.Wait()
	defer cmd.Process.Kill()

	body, err := waitForBody(fmt.Sprintf("http://127.0.0.1:%d%s", data.port, metricPath))
	if err != nil {
		t.Fatalf("unable to get metrics: %s", err)
	}
	got := string(body)

	metricsPrefixes := []string{
		"go_gc_duration_seconds",
		"go_goroutines",
		"go_memstats",
	}

	for _, prefix := range metricsPrefixes {
		if !strings.Contains(got, prefix) {
			t.Fatalf("no metric starting with %s", prefix)
		}
	}
}

func testBuildVersionGatherer(t *testing.T, data bin) {
	metricPath := "/metrics"
	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()

	cmd := exec.CommandContext(
		ctx,
		data.path,
		"--web.telemetry-path", metricPath,
		"--web.listen-address", fmt.Sprintf(":%d", data.port),
	)

	if err := cmd.Start(); err != nil {
		t.Fatal(err)
	}
	defer cmd.Wait()
	defer cmd.Process.Kill()

	body, err := waitForBody(fmt.Sprintf("http://127.0.0.1:%d%s", data.port, metricPath))
	if err != nil {
		t.Fatalf("unable to get metrics: %s", err)
	}
	got := string(body)

	metricsPrefixes := []string{
		"mongodb_exporter_build_info",
	}

	for _, prefix := range metricsPrefixes {
		if !strings.Contains(got, prefix) {
			t.Fatalf("no metric starting with %s", prefix)
		}
	}
}

func testFlagTest(t *testing.T, data bin) {
	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()

	cmd := exec.CommandContext(
		ctx,
		data.path,
		"--test",
	)

	b, err := cmd.CombinedOutput()
	if err != nil {
		t.Log(string(b))
		t.Fatal(err)
	}
	buildInfo := shared.BuildInfo{}
	err = json.Unmarshal(b, &buildInfo)
	if err != nil {
		t.Fatal(err)
	}

	if reflect.DeepEqual(buildInfo, shared.BuildInfo{}) {
		t.Fatalf("buildInfo is empty")
	}
}

func testFlagTestWithTLS(t *testing.T, data bin) {
	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()

	cmd := exec.CommandContext(
		ctx,
		data.path,
		"--mongodb.tls",
		"--mongodb.tls-ca=testdata/ca.crt",
		"--mongodb.tls-cert=testdata/client.pem",
		"--test",
	)

	b, err := cmd.CombinedOutput()
	if err != nil {
		t.Log(string(b))
		t.Fatal(err)
	}
	buildInfo := shared.BuildInfo{}
	err = json.Unmarshal(b, &buildInfo)
	if err != nil {
		t.Fatal(err)
	}

	if reflect.DeepEqual(buildInfo, shared.BuildInfo{}) {
		t.Fatalf("buildInfo is empty")
	}
}

// waitForBody is a helper function which makes http calls until http server is up
// and then returns body of the successful call.
func waitForBody(urlToGet string) (body []byte, err error) {
	tries := 60

	// Get data, but we need to wait a bit for http server.
	for i := 0; i <= tries; i++ {
		// Try to get web page.
		body, err = getBody(urlToGet)
		if err == nil {
			return body, err
		}

		// If there is a syscall.ECONNREFUSED error (web server not available) then retry.
		if urlError, ok := err.(*url.Error); ok {
			if opError, ok := urlError.Err.(*net.OpError); ok {
				if osSyscallError, ok := opError.Err.(*os.SyscallError); ok {
					if osSyscallError.Err == syscall.ECONNREFUSED {
						time.Sleep(1 * time.Second)
						continue
					}
				}
			}
		}

		// There was an error, and it wasn't syscall.ECONNREFUSED.
		return nil, err
	}

	return nil, fmt.Errorf("failed to GET %s after %d tries: %s", urlToGet, tries, err)
}

// getBody is a helper function which retrieves http body from given address.
func getBody(urlToGet string) ([]byte, error) {
	resp, err := http.Get(urlToGet)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	return body, nil
}

func TestVersionInfo(t *testing.T) {
	t.Run("Check version parameters", func(t *testing.T) {
		var currentTime = time.Now()
		pmmVersion.Version = "1.2.3"
		pmmVersion.PMMVersion = "4.5.6"
		pmmVersion.FullCommit = "test-commit-sha"
		pmmVersion.Branch = "test-branch"
		pmmVersion.Timestamp = strconv.FormatInt(currentTime.Unix(), 10)

		initVersionInfo()

		assert.Equal(t, pmmVersion.Version+"-pmm-"+pmmVersion.PMMVersion, version.Version)
		assert.Equal(t, pmmVersion.FullCommit, version.Revision)
		assert.Equal(t, pmmVersion.Branch, version.Branch)
		assert.Equal(t, currentTime.Format(versionDataFormat), version.BuildDate)
	})

	t.Run("Check Empty Timestamp", func(t *testing.T) {
		pmmVersion.Timestamp = ""
		initVersionInfo()
		assert.Equal(t, time.Unix(0, 0).Format(versionDataFormat), version.BuildDate)
	})

	t.Run("Check PMMVersion Empty", func(t *testing.T) {
		pmmVersion.Version = "1.2.3"
		pmmVersion.PMMVersion = ""
		initVersionInfo()
		assert.Equal(t, pmmVersion.Version, version.Version)
	})
}


mongodb_exporter.go
// Copyright 2017 Percona LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"fmt"
	"os"
	"strconv"
	"time"

	"github.com/percona/exporter_shared"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/prometheus/common/log"
	"github.com/prometheus/common/version"
	"gopkg.in/alecthomas/kingpin.v2"

	pmmVersion "github.com/percona/pmm/version"

	"github.com/ghostbaby/mongodb_exporter/collector"
	"github.com/ghostbaby/mongodb_exporter/shared"
)

const (
	program           = "mongodb_exporter"
	versionDataFormat = "20060102-15:04:05"
)

var (
	listenAddressF = kingpin.Flag("web.listen-address", "Address to listen on for web interface and telemetry.").Default(":9216").String()
	metricsPathF   = kingpin.Flag("web.telemetry-path", "Path under which to expose metrics.").Default("/metrics").String()

	collectDatabaseF   = kingpin.Flag("collect.database", "Enable collection of Database metrics").Bool()
	collectCollectionF = kingpin.Flag("collect.collection", "Enable collection of Collection metrics").Bool()
	collectTopF        = kingpin.Flag("collect.topmetrics", "Enable collection of table top metrics").Bool()
	collectIndexUsageF = kingpin.Flag("collect.indexusage", "Enable collection of per index usage stats").Bool()

	uriF = kingpin.Flag("mongodb.uri", "MongoDB URI, format").
		PlaceHolder("[mongodb://][user:pass@]host1[:port1][,host2[:port2],...][/database][?options]").
		Default("mongodb://localhost:27017").
		Envar("MONGODB_URI").
		String()

	authDB   = kingpin.Flag("mongodb.authentification-database", "Specifies the database in which the user is created").Default("").String()
	tlsF     = kingpin.Flag("mongodb.tls", "Enable tls connection with mongo server").Bool()
	tlsCertF = kingpin.Flag("mongodb.tls-cert", "Path to PEM file that contains the certificate (and optionally also the decrypted private key in PEM format).\n"+
		"    \tThis should include the whole certificate chain.\n"+
		"    \tIf provided: The connection will be opened via TLS to the MongoDB server.").Default("").String()
	tlsPrivateKeyF = kingpin.Flag("mongodb.tls-private-key", "Path to PEM file that contains the decrypted private key (if not contained in mongodb.tls-cert file).").Default("").String()
	tlsCAF         = kingpin.Flag("mongodb.tls-ca", "Path to PEM file that contains the CAs that are trusted for server connections.\n"+
		"    \tIf provided: MongoDB servers connecting to should present a certificate signed by one of this CAs.\n"+
		"    \tIf not provided: System default CAs are used.").Default("").String()
	tlsDisableHostnameValidationF = kingpin.Flag("mongodb.tls-disable-hostname-validation", "Disable hostname validation for server connection.").Bool()
	maxConnectionsF               = kingpin.Flag("mongodb.max-connections", "Max number of pooled connections to the database.").Default("1").Int()
	testF                         = kingpin.Flag("test", "Check MongoDB connection, print buildInfo() information and exit.").Bool()

	socketTimeoutF = kingpin.Flag("mongodb.socket-timeout", "Amount of time to wait for a non-responding socket to the database before it is forcefully closed.\n"+
		"    \tValid time units are 'ns', 'us' (or 'µs'), 'ms', 's', 'm', 'h'.").Default("3s").Duration()
	syncTimeoutF = kingpin.Flag("mongodb.sync-timeout", "Amount of time an operation with this session will wait before returning an error in case\n"+
		"    \ta connection to a usable server can't be established.\n"+
		"    \tValid time units are 'ns', 'us' (or 'µs'), 'ms', 's', 'm', 'h'.").Default("1m").Duration()

	// FIXME currently ignored
	// enabledGroupsFlag = flag.String("groups.enabled", "asserts,durability,background_flushing,connections,extra_info,global_lock,index_counters,network,op_counters,op_counters_repl,memory,locks,metrics", "Comma-separated list of groups to use, for more info see: docs.mongodb.org/manual/reference/command/serverStatus/")
	enabledGroupsFlag = kingpin.Flag("groups.enabled", "Currently ignored").Default("").String()
)

func main() {
	initVersionInfo()
	log.AddFlags(kingpin.CommandLine)
	kingpin.Parse()

	if *testF {
		buildInfo, err := shared.TestConnection(
			shared.MongoSessionOpts{
				URI:                   *uriF,
				TLSConnection:         *tlsF,
				TLSCertificateFile:    *tlsCertF,
				TLSPrivateKeyFile:     *tlsPrivateKeyF,
				TLSCaFile:             *tlsCAF,
				TLSHostnameValidation: !(*tlsDisableHostnameValidationF),
				AuthentificationDB:    *authDB,
			},
		)
		if err != nil {
			log.Errorf("Can't connect to MongoDB: %s", err)
			os.Exit(1)
		}
		fmt.Println(string(buildInfo))

		//ServerStatus := mongod.TestServerStatus(
		//	mongod.MongoSessionOpts{
		//		URI:                   *uriF,
		//		TLSConnection:         *tlsF,
		//		TLSCertificateFile:    *tlsCertF,
		//		TLSPrivateKeyFile:     *tlsPrivateKeyF,
		//		TLSCaFile:             *tlsCAF,
		//		TLSHostnameValidation: !(*tlsDisableHostnameValidationF),
		//		AuthentificationDB:    *authDB,
		//	},
		//)
		//if ServerStatus != nil {
		//	log.Errorf("Can't get MongoDB Server Status: %s", ServerStatus)
		//	os.Exit(1)
		//}
		os.Exit(0)
	}

	// TODO: Maybe we should move version.Info() and version.BuildContext() to https://github.com/percona/exporter_shared
	// See: https://jira.percona.com/browse/PMM-3250 and https://github.com/ghostbaby/mongodb_exporter/pull/132#discussion_r262227248
	log.Infoln("Starting", program, version.Info())
	log.Infoln("Build context", version.BuildContext())

	programCollector := version.NewCollector(program)
	mongodbCollector := collector.NewMongodbCollector(&collector.MongodbCollectorOpts{
		URI:                      *uriF,
		TLSConnection:            *tlsF,
		TLSCertificateFile:       *tlsCertF,
		TLSPrivateKeyFile:        *tlsPrivateKeyF,
		TLSCaFile:                *tlsCAF,
		TLSHostnameValidation:    !(*tlsDisableHostnameValidationF),
		DBPoolLimit:              *maxConnectionsF,
		CollectDatabaseMetrics:   *collectDatabaseF,
		CollectCollectionMetrics: *collectCollectionF,
		CollectTopMetrics:        *collectTopF,
		CollectIndexUsageStats:   *collectIndexUsageF,
		SocketTimeout:            *socketTimeoutF,
		SyncTimeout:              *syncTimeoutF,
		AuthentificationDB:       *authDB,
	})
	prometheus.MustRegister(programCollector, mongodbCollector)

	promHandler := promhttp.InstrumentMetricHandler(prometheus.DefaultRegisterer, promhttp.HandlerFor(prometheus.DefaultGatherer, promhttp.HandlerOpts{ErrorHandling: promhttp.ContinueOnError}))
	exporter_shared.RunServer("MongoDB", *listenAddressF, *metricsPathF, promHandler)
}

// initVersionInfo sets version info
// If binary was build for PMM with environment variable PMM_RELEASE_VERSION
// `--version` will be displayed in PMM format. Also `PMM Version` will be connected
// to application version and will be printed in all logs.
// TODO: Refactor after moving version.Info() and version.BuildContext() to https://github.com/percona/exporter_shared
// See: https://jira.percona.com/browse/PMM-3250 and https://github.com/ghostbaby/mongodb_exporter/pull/132#discussion_r262227248
func initVersionInfo() {
	version.Version = pmmVersion.Version
	version.Revision = pmmVersion.FullCommit
	version.Branch = pmmVersion.Branch

	if buildDate, err := strconv.ParseInt(pmmVersion.Timestamp, 10, 64); err != nil {
		version.BuildDate = time.Unix(0, 0).Format(versionDataFormat)
	} else {
		version.BuildDate = time.Unix(buildDate, 0).Format(versionDataFormat)
	}

	if pmmVersion.PMMVersion != "" {
		version.Version += "-pmm-" + pmmVersion.PMMVersion
		kingpin.Version(pmmVersion.FullInfo())
	} else {
		kingpin.Version(version.Print(program))
	}

	kingpin.HelpFlag.Short('h')
	kingpin.CommandLine.Help = fmt.Sprintf("%s exports various MongoDB metrics in Prometheus format.\n", pmmVersion.ShortInfo())
}
